ehsm_kms_service/function.js

```diff
@@ -444,71 +450,91 @@ const _checkParams = function (req, res, next, nonce_database, DB) {
-    } else if (
-      !!nonce_database[appid] &&
-      nonce_database[appid].findIndex(
-        (nonce_data) => nonce_data.nonce == nonce
-      ) > -1
-    ) {
-      res.send(_result(400, "Timestamp can't be repeated in 20 minutes"))
-      return
-    } else {
-      nonce_database[appid].unshift(nonce_data)
-    }
 
-    if (ACTION !== key_management_apis[ACTION]) {
-      // check payload
-      const _checkPayload_res = _checkPayload(req, res, next)
-      if (!_checkPayload_res) {
-        return
+    var _lastSteps = () => {
+      if (ACTION !== key_management_apis[ACTION]) {
+        // check payload
+        const _checkPayload_res = _checkPayload(req, res, next)
+        if (!_checkPayload_res) {
+          return
+        }
       }
-    }
 
-    // check sign
-    let sign_params = { appid, timestamp, payload }
+      // check sign
+      let sign_params = { appid, timestamp, payload }
 
-    gen_hmac(DB, appid, sign_params)
-      .then(result => {
-        if (result.hmac.length == 0) {
-          res.send(_result(400, result.error))
-          return
-        }
-        if (sign != result.hmac ) {
-          res.send(_result(400, 'sign error'))
-          return
+      gen_hmac(DB, appid, sign_params)
+        .then(result => {
+          if (result.hmac.length == 0) {
+            ReportControlTarget(req.ip, new Date().getTime())
+            res.send(_result(400, result.error))
+            return
+          }
+          if (sign != result.hmac ) {
+            ReportControlTarget(req.ip, new Date().getTime())
+            res.send(_result(400, 'sign error'))
+            return
+          } else {
+            next()
+          }
+        })
+        .catch((e) => {
+          res.send(_result(400, 'database error'))
+        })
+    }
+
+    var _checkNonce = CheckNonce(nonce_database, appid, nonce, new Date().getTime())
+    if (_checkNonce == undefined) {
+      /**
+       * Cache nonce locally after receiving the request
+       * nonce_database - object
+       *  {
+       *    <appid>: [nonce_data，nonce_data]
+       *  }
+       * nonce_data - object
+       *  {
+       *    nonce: ****
+       *    nonce_timestamp: new Date().getTime()
+       *  }
+       */
+      const nonce_data = { nonce, nonce_timestamp: new Date().getTime() }
+      if (!nonce_database[appid]) {
+        nonce_database[appid] = [nonce_data]
+      } else if (
+        !!nonce_database[appid] &&
+        nonce_database[appid].findIndex(
+          (nonce_data) => nonce_data.nonce == nonce
+        ) > -1
+      ) {
+        res.send(_result(400, "Timestamp can't be repeated in 20 minutes"))
+        return
+      } else {
+        nonce_database[appid].unshift(nonce_data)
+      }
+      _lastSteps()
+    } else {
+      _checkNonce.then((checkNonceRes) => {
+        if (checkNonceRes) {
+          _lastSteps()
         } else {
-          next()
+          res.send(_result(400, "Timestamp can't be repeated in 20 minutes"))

```

Comment:
ReportControlTarget(req.ip, new Date().getTime())

Reply:
This is same as before.